### PR TITLE
[Gst/MQTT/Src] Keep subscribing to the topic only in the PLAYING state

### PR DESCRIPTION
**Pull request by wooksong**
This patch makes the subscription be removed when the state of the element is changed to PAUSED from PLAYING. In addition, to block the extra invocation of the receiving message handler running on the other thread, checking the subscription flag is added to this handler.

Signed-off-by: Wook Song <wook16.song@samsung.com>

**Changelog**
- Handled the case when the state is changed to PLAYING again from PAUSED (https://github.com/nnstreamer/nnstreamer/pull/3279#pullrequestreview-661585524)

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/mqtt/mqttsrc.c [hunk 1]
```diff
@@ -445,6 +445,20 @@ gst_mqtt_src_change_state (GstElement * element, GstStateChange transition)
...
+        int sub_ret;
+
+        sub_ret = MQTTAsync_unsubscribe (self->mqtt_client_handle,
+            self->mqtt_topic, &self->mqtt_respn_opts);
+        if (sub_ret != MQTTASYNC_SUCCESS) {
+          GST_ERROR_OBJECT (self,
+              "MQTTAsync_unsubscribe failed with return code %d", sub_ret);
+        }
+
+        self->is_subscribed = FALSE;
```
Comment (dongju-chae) [inline gst/mqtt/mqttsrc.c:459]: Even if `MQTTAsync_unsubscribe` is failed?

Comment (wooksong) [reply gst/mqtt/mqttsrc.c:459]: > Even if `MQTTAsync_unsubscribe` is failed?

At this point, there is nothing we can do.

Comment (dongju-chae) [reply gst/mqtt/mqttsrc.c:459]: true..

**Review by dongju-chae — APPROVED** [review]
LGTM

**Review by helloahn — APPROVED** [review]
LGTM

**Review by myungjoo — COMMENTED** [review]
Do you have re-subscribe mechanism for PAUSE-->PLAYING?

(Tested with start/plyaing --> pause --> resume/playing --> pause --> ... ?)
You'll need corresponding UTC (verifying it's actually "ubsub-ed" for pause, too) for this as well.

**Comment by wooksong** [comment]
> Do you have re-subscribe mechanism for PAUSE-->PLAYING?
> 
> (Tested with start/plyaing --> pause --> resume/playing --> pause --> ... ?)
> You'll need corresponding UTC (verifying it's actually "ubsub-ed" for pause, too) for this as well.

As [this figure](https://gstreamer.freedesktop.org/documentation/application-development/advanced/images/clocks.png) represented, the state of the pipeline only affects the stream time. We are using the clock time so that timestamp conversion and synchronization would be OK, IMO. 
- Tested with sub/unsub and sub again by terminating while running the whole demonstration scenario. 
- Not (Tested with start/plyaing --> pause --> resume/playing --> pause --> ... ?)

File: gst/mqtt/mqttsrc.c [hunk 2]
```diff
@@ -437,7 +439,14 @@ gst_mqtt_src_change_state (GstElement * element, GstStateChange transition)
       diff = GST_CLOCK_DIFF (base_time, cur_time);
       self->base_time_epoch =
           g_get_real_time () * GST_US_TO_NS_MULTIPLIER - diff;
-      GST_INFO_OBJECT (self, "GST_STATE_CHANGE_PAUSED_TO_PLAYING");
+
+      /** This handles the case when the state is changed to PLAYING again */
+      if (self->is_connected && !_do_subscribe (self)) {
+        GST_ERROR_OBJECT (self, "Failed to re-subscribe to %s",
+            self->mqtt_topic);
```
Comment (myungjoo) [inline gst/mqtt/mqttsrc.c:446]: sub/unsub methods don't look symmetric and touch different properties. Can you improve it?

1. this uses is_connected while the other uses is_subscribed
2. this uses an abstracted func, "_do_subsribe", while the other uses MQTT API directly.
3. this do not update any properties directly, while the other do so.

Recommendation: use the same property to check status and make _do_unsubscribe that is symmetric to _do_subscribe.

Not using mutex in this (while the does does) is also a bit itchy.

**Review by myungjoo — APPROVED** [review]
LGTM

**Pull request merged**